nabs/optimize.py

```diff
@@ -0,0 +1,289 @@
+import math
+import logging
+
+import bluesky.preprocessors as bpp
+import bluesky.plan_stubs as bps
+from scipy.constants import golden_ratio
+
+from nabs.plan_stubs import measure_average
+from nabs.streams import AverageStream
+from nabs.utils import InvertedSignal, ErrorSignal
+
+logger = logging.getLogger(__name__)
+
+
+def minimize(*args, **kwargs):
+    """
+    Minimize the value of an ophyd.Signal
+    """
+    # Add plan name into metadata
+    _md = {'plan_name': 'minimize'}
+    _md.update(kwargs.get('md', {}))
+    # Run the plan
+    return (yield from optimize(*args, maximize=False, md=_md, **kwargs))
+
+
+def maximize(*args, **kwargs):
+    """
+    Maximize the value of an ophyd.Signal
+    """
+    # Add plan name into metadata
+    _md = {'plan_name': 'maximize'}
+    _md.update(kwargs.get('md', {}))
+    # Run the plan
+    return (yield from optimize(*args, maximize=True, **kwargs))
+
+
+def walk_to_target(signal, motor, target, tolerance, **kwargs):
+    """
+    Walk the motor until a signal reaches our target
+
+    Similar to the :func:`.maximize` and :func:`.minimize`. There are options
+    for multiple algorithms to dictate the scanning procedure. This may change
+    the interpretation of values passed into this scanning procedure.
+
+    Parameters
+    ----------
+    signal: ophyd.Signal
+        Signal to maximize
+
+    motor: ophyd.OphydObject
+        Any set-able object
+
+    tolerance: float, optional
+        The tolerance in which our motor is
+
+    average: int, optional
+        Choice to take an average of points at each point in the scan
+
+    limits: tuple, optional
+        Limit the region the scan will search within. If this is not provided,
+        the soft limits of the signal will be used. In this case, these must be
+        configured or the scan will not be allowed to continue.
+
+    method: str, optional
+        Choice of optimization methods
+
+    md: dict, optional
+        metadata
+    """
+    # Add walk information to metadata
+    _md = {'plan_name': 'walk_to_target',
+           'target': target}
+    _md.update(kwargs.get('md', {}))
+    # Create a signal whose value is the absolute error
+    error = ErrorSignal(signal, target)
+    return (yield from minimize(error, motor, tolerance, **kwargs))
+
+
+def optimize(signal, motor, tolerance,
+             average=None, limits=None, method='golden',
+             maximize=False, md=None):
+    """
+    The optimization methods within ``nabs`` take a similar approach to the
```

Comment:
Imagine making an A+ pull request and forgetting the 1-line doc summary as the first line convention

Reply:
This is because I add it to the other functions doc-strings and didn't want two 1-line doc summaries. You think this is a good thing to do or should I just bite the bullet and duplicate the docstring?

Reply:
If this function is included in the sphinx docs it should follow the convention, otherwise it doesn't matter. You can always include this `__doc__` outside of the function and append it to all functions that need it, and have just a one line summary here that gets added to.

Reply:
Yeah. I think that sounds like a better option. I'll play around with this.